pkgs/development/libraries/science/math/tensorrt/generic.nix

```diff
@@ -0,0 +1,65 @@
+{ lib
+, stdenv
+, callPackage
+, autoPatchelfHook
+, autoAddOpenGLRunpathHook
+, cudatoolkit
+, cudnn
+
+, version
+}:
+
+# Ensure the passed version is one that we know how to handle.
+assert (lib.assertOneOf "version" version (map (v: v.version) (callPackage ./versions.nix { })));
+
+let
+  mostOfVersion = builtins.concatStringsSep "."
+    (lib.take 3 (lib.versions.splitVersion version));
+  foundVersion = lib.lists.findFirst (v: v.version == version) "none" (callPackage ./versions.nix { });
+in
+stdenv.mkDerivation {
+  pname = "cudatoolkit-${cudatoolkit.majorVersion}-tensorrt";
+  inherit version;
+  inherit (foundVersion) src;
+
+  outputs = [ "out" "dev" ];
+
+  nativeBuildInputs = [
+    autoPatchelfHook
+    autoAddOpenGLRunpathHook
+  ];
+
+  # Used by autoPatchelfHook
+  buildInputs = [
+    cudatoolkit.cc.cc.lib # libstdc++
+    cudatoolkit
+    cudnn
+  ];
+
+  sourceRoot = "TensorRT-${version}";
+
+  installPhase = ''
+    install --directory "$dev" "$out"
+    mv include "$dev"
+    mv targets/x86_64-linux-gnu/lib "$out"
+    install -D --target-directory="$out/bin" targets/x86_64-linux-gnu/bin/trtexec
+  '';
+
+  # Make sure the shared-object files within this package can find each other at
+  # runtime.
+  postFixup = ''
+    echo 'Patching RPATH of libnvinfer libs'
+    patchelf --debug --add-rpath '$ORIGIN' \
```

Comment:
This seems to execute before `autoPatchelfHook`, s.t. autoPatchelf removes `$ORIGIN` for the final rpath.
You can verify that by building `cudaPackages.tensorrt` and checking out `patchelf --print-rpath ./result/lib/libnvinfer*`
The test you suggested before (`assert tensorrt.Builder(tensorrt.Logger())`) fails looking for `libnvinfer.*` things.

It would be nice to have a way to solve this without directly using `patchelf`, but a solution that'll work for now is to modify not the `rpath`, but the list of "needed" dependencies so that `autoPatchelf` knows which paths to keep:

```
  # Inform autoPatchelf to keep $ORIGIN
  postFixup = ''
    echo 'Patching NEEDED of libnvinfer libs'
    patchelf --debug --add-needed 'libnvinfer.so' \
      "$out/lib/libnvinfer.so.${mostOfVersion}" \
      "$out/lib/libnvinfer_plugin.so.${mostOfVersion}" \
      "$out/lib/libnvinfer_builder_resource.so.${mostOfVersion}"
  '';
```